src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,585 @@
+#include "aggregate_sort.hpp"
+
+#include <memory>
+#include <vector>
+
+#include "aggregate/aggregate_traits.hpp"
+#include "all_type_variant.hpp"
+#include "constant_mappings.hpp"
+#include "operators/sort.hpp"
+#include "storage/segment_iterate.hpp"
+#include "table_wrapper.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+AggregateSort::AggregateSort(const std::shared_ptr<AbstractOperator>& in,
+                             const std::vector<AggregateColumnDefinition>& aggregates,
+                             const std::vector<ColumnID>& groupby_column_ids)
+    : AbstractAggregateOperator(in, aggregates, groupby_column_ids) {}
+
+const std::string AggregateSort::name() const { return "AggregateSort"; }
+
+/**
+ * Calculates the value for the <code>aggregate_index</code>'th aggregate.
+ * To do this, we iterate over all segments of the corresponding column.
+ * The aggregate values (one per group-by-combination) are built incrementally.
+ * Every time we reach the begin of a new group-by-combination,
+ * we store the aggregate value for the current (now previous) group.
+ *
+ * @tparam ColumnType the type of the input column to aggregate on
+ * @tparam AggregateType the type of the aggregate (=output column)
+ * @tparam function the aggregate function, as type parameter - e.g. AggregateFunction::MIN, AVG, COUNT, ...
+ * @param aggregate_group_pointers the row ids where a new combination in the sorted table begins
+ * @param aggregate_index determines which aggregate to calculate (from _aggregates)
+ * @param aggregate_function the aggregate function - as callable object. This should always be the same as <code>function</code>
+ * @param sorted_table the input table, sorted by the group by columns
+ */
+template <typename ColumnType, typename AggregateType, AggregateFunction function>
+void AggregateSort::_aggregate_values(const std::set<RowID>& aggregate_group_pointers, const uint64_t aggregate_index,
+                                      const std::shared_ptr<const Table>& sorted_table) {
+  auto aggregate_function = AggregateFunctionBuilder<ColumnType, AggregateType, function>().get_aggregate_function();
+
+  // We already know beforehand how many aggregate values (=group-by-combinations) we have to calculate
+  const size_t num_groups = aggregate_group_pointers.size() + 1;
+
+  // Vectors to store aggregate values (and if they are NULL) for later usage in value segments
```

Comment:
Why aren't these concurrent vectors that can directly be moved (not copied) into the value segments?

Reply:
As far as I understood, the `pmr_concurrent_vector` performs some synchronization, we (currently) don't need. But I am not sure, how big the impact is compared to an extra copy?

Reply:
It's not only about synchronization which comes with TBB's concurrent vectors but also about data placement, i.e., where (on which NUMA node or on DRAM/NVRAM) memory is allocated.

Regarding the impact: If we can, we want to avoid extra copies.

Reply:
In the [jit aggregate](https://github.com/hyrise/hyrise/blob/master/src/lib/operators/jit_operator/operators/jit_aggregate.cpp) is a similar situation. There is [a comment](https://github.com/hyrise/hyrise/blob/master/src/lib/operators/jit_operator/operators/jit_aggregate.cpp#L109) that argues that the std vector (including copying) is (in that case) faster than using the `pmr_concurrent_vector`?

Reply:
I actually never measured the difference in performance in creating an `std::vector<whatever>` first and inserting it into a `concurrent_vector` compared to that of constructing the latter. As the `concurrent_vector` is due to die anyway (#749 - hihi, I just have to type `#getrid`), let's leave it the way it is.